sdks/java/io/distributedlog/src/main/java/org/apache/beam/sdk/io/distributedlog/DLBoundedSource.java

```diff
@@ -0,0 +1,192 @@
+      Optional<URI> confUri,
+      List<String> streams,
+      Coder<R> rCoder,
+      LogSegmentBundle segment) {
+    this.dlUri = dlUri;
+    this.confUri = confUri;
+    this.streams = streams;
+    this.rCoder = rCoder;
+    this.segment = segment;
+  }
+
+  DistributedLogNamespace setupNamespace() throws IOException {
+    DistributedLogConfiguration conf = new DistributedLogConfiguration();
+    if (confUri.isPresent()) {
+      try {
+        conf.loadConf(confUri.get().toURL());
+      } catch (ConfigurationException e) {
+        throw new IOException("Failed to load distributedlog configuration from " + confUri, e);
+      } catch (MalformedURLException e) {
+        throw new IOException("Invalid distributedlog configuration uri " + confUri, e);
+      }
+    }
+    return DistributedLogNamespaceBuilder.newBuilder()
+        .conf(conf)
+        .uri(dlUri)
+        .clientId("distributedlog-bounded-source-" + InetAddress.getLocalHost().getHostName())
+        .statsLogger(NullStatsLogger.INSTANCE)
+        .regionId(DistributedLogConstants.LOCAL_REGION_ID)
+        .build();
+  }
+
+  List<String> getStreams() {
+    return streams;
+  }
+
+  @Override
+  public void validate() {
+    checkArgument(!streams.isEmpty(), "need to set the streams of a DLBoundedSource");
+    checkNotNull(rCoder, "need to set the key coder of a DLBoundedSource");
+  }
+
+  static List<LogSegmentBundle> getAllLogSegments(
+      DistributedLogNamespace namespace,
+      List<String> streams) throws IOException {
+    List<LogSegmentBundle> segments = Lists.newArrayList();
+    for (String stream : streams) {
+      segments.addAll(getLogSegments(namespace, stream));
+    }
+    return segments;
+  }
+
+  private static List<LogSegmentBundle> getLogSegments(
+      DistributedLogNamespace namespace,
+      final String stream) throws IOException {
+    DistributedLogManager manager = namespace.openLog(stream);
+    try {
+      return Lists.transform(manager.getLogSegments(),
+          new Function<LogSegmentMetadata, LogSegmentBundle>() {
+            @Nullable
+            @Override
+            public LogSegmentBundle apply(@Nullable LogSegmentMetadata metadata) {
+              return new LogSegmentBundle(stream, metadata);
+            }
+          });
+    } finally {
+      manager.close();
+    }
+  }
+
+  @Override
+  public List<? extends BoundedSource<R>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                           PipelineOptions options) throws Exception {
+    DistributedLogNamespace namespace = setupNamespace();
+    try {
+      if (null == segment) {
+        return Lists.transform(getAllLogSegments(namespace, streams),
+            new Function<LogSegmentBundle, BoundedSource<R>>() {
+              @Override
+              public BoundedSource<R> apply(@Nullable LogSegmentBundle bundle) {
+                return new DLBoundedSource<>(
+                    dlUri,
+                    confUri,
```

Comment:
The split is based on the bundle right ? It means that we will have one source per bundle ?